src/Symfony/Component/Validator/Constraints/UlidValidator.php

```diff
@@ -40,16 +40,28 @@ public function validate(mixed $value, Constraint $constraint): void
 
         $value = (string) $value;
 
-        if (26 !== \strlen($value)) {
+        switch ($constraint->format) {
+            case Ulid::FORMAT_BASE_32:
+                $requiredLength = 26;
+                $requiredCharset = '0123456789ABCDEFGHJKMNPQRSTVWXYZabcdefghjkmnpqrstvwxyz';
+                break;
+            case Ulid::FORMAT_BASE_58:
+                $requiredLength = 22;
+                $requiredCharset = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
+                break;
+                // no default, the constraint constructor already ensures a valid format
+        }
+
+        if ($requiredLength !== \strlen($value)) {
             $this->context->buildViolation($constraint->message)
                 ->setParameter('{{ value }}', $this->formatValue($value))
-                ->setCode(26 > \strlen($value) ? Ulid::TOO_SHORT_ERROR : Ulid::TOO_LONG_ERROR)
+                ->setCode($requiredLength > \strlen($value) ? Ulid::TOO_SHORT_ERROR : Ulid::TOO_LONG_ERROR)
                 ->addViolation();
 
             return;
         }
 
-        if (\strlen($value) !== strspn($value, '0123456789ABCDEFGHJKMNPQRSTVWXYZabcdefghjkmnpqrstvwxyz')) {
+        if (\strlen($value) !== strspn($value, $requiredCharset)) {
```

Comment:
Should we add the format used to validate as a parameter to the violation as well? This way users can reuse it in their custom validation error messages.

Reply:
Totally agree